### PR TITLE
RATEST-279: Fix test to run on firefox browser

### PR DESCRIPTION
**Pull request by Irenyak1**
Ticket ->  [RATEST-279](https://issues.openmrs.org/browse/RATEST-279)

Fixed -> End Visit User story to run on Firefox engine

### CODE REVIEW

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java [hunk 1]
```diff
@@ -81,37 +80,6 @@ public void launchPatientDashboardPage() {
 		assertNotNull(dashboardPage.getActiveVisitList());
 	}
```
Comment (Irenyak1) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java:82]: For the sake of uniformity I aligned the steps definition file to match the feature file steps.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java [hunk 2]
```diff
@@ -81,37 +80,6 @@ public void launchPatientDashboardPage() {
...
-		visitNotePage.save();
-	}
-
-	@Then("the system adds the note into visit note table")
-	public void systemAddsVisitNote() {
-		assertEquals(DIAGNOSIS_PRIMARY, visitNotePage.primaryDiagnosis());
-		assertEquals(DIAGNOSIS_SECONDARY, visitNotePage.secondaryDiagnosis());
-		visitsDashboardPage.waitForPage();
-	}
-
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java:114]: How is this change related to what you are fixing?

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/PatientVisitsDashboardPage.java [hunk 3]
```diff
@@ -74,8 +74,9 @@ public void deleteVitals() {
 		return findElements(VITALS_LIST);
 	}
 
-	public void goToVisitNote() {
-		findElement(VISIT_NOTE).click();
+	public VisitNotePage goToVisitNote() {
+		clickOnLast(VISIT_NOTE);
+		return new VisitNotePage(this);
```
Comment (kdaud) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/PatientVisitsDashboardPage.java:79]: How is this change related to what you are fixing?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java [hunk 4]
```diff
@@ -81,37 +80,6 @@ public void launchPatientDashboardPage() {
...
-		assertTrue(textExists("Visit Note"));
-	}
-
-	@When("a user fills the visit note")
-	public void fillVisitNoteForm() {
-		visitNotePage.selectProviderAndLocation();
-		visitNotePage.addDiagnosis(DIAGNOSIS_PRIMARY);
-		visitNotePage.addSecondaryDiagnosis(DIAGNOSIS_SECONDARY);
-		visitNotePage.addNote("This is a visit note.");
-	}
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java:101]: How is this change related to what you are fixing?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java [hunk 5]
```diff
@@ -81,37 +80,6 @@ public void launchPatientDashboardPage() {
...
-	public void loadVisitNotePage() {
-		visitsDashboardPage = dashboardPage.goToRecentVisits();
-		visitNotePage = (VisitNotePage) dashboardPage.goToVisitNote().waitForPage();
-	}
-
-	@Then("the system loads visit note page")
-	public void systemLoadsVisitNotePage() {
-		assertTrue(textExists("Visit Note"));
-	}
-
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java:94]: To add on @kdaud  comment, Fix what is required

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/PatientVisitsDashboardPage.java (hunk 3 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/PatientVisitsDashboardPage.java:79]: @kdaud , @sherrif10 the reason I implemented that method in this page was to cater for this step. 

` @When("a user clicks visit note link from the patient dashboard")
	public void loadVisitNotePage() {
		visitsDashboardPage = dashboardPage.goToRecentVisits();
		visitNotePage = (VisitNotePage) dashboardPage.goToVisitNote().waitForPage();
	}  `

so that it could be written like this 

` @When("a user clicks visit note link from the patient dashboard")
	public void loadVisitNotePage() {
		visitsDashboardPage = dashboardPage.goToRecentVisits();
		visitNotePage = (VisitNotePage) visitsDashboardPage.goToVisitNote().waitForPage();
	}  `

why would the test go to visitsDashboardPage and there after call goToVisitNote() from the dashboardPage, why not call it from visitsDashboardPage? What do you guys think?

Comment (kdaud) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/PatientVisitsDashboardPage.java:79]: > why would the test go to visitsDashboardPage and there after call goToVisitNote() from the dashboardPage

ClinicianFacingdashboardPage is a parent page to VisitNotePage so a child page can access the functionality from the parent page as long as the functionality exists on the parent page.

> why not call it from visitsDashboardPage?

Yes the functionality can also be invoked directly on the child page.

Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/PatientVisitsDashboardPage.java:79]: @kdaud so what can I implement in this case?

Comment (kdaud) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/PatientVisitsDashboardPage.java:79]: @Irenyak1 is the existing implementation failing? And is the change related to what the ticket is addressing?

Comment (Irenyak1) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/PatientVisitsDashboardPage.java:79]: @kdaud the implementation is not failing but there's code redundancy in this step
`@When("a user clicks visit note link from the patient dashboard") 
public void loadVisitNotePage() { 
visitsDashboardPage = dashboardPage.goToRecentVisits(); 
visitNotePage = (VisitNotePage) dashboardPage.goToVisitNote().waitForPage(); 
}`

Comment (kdaud) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/PatientVisitsDashboardPage.java:79]: > @kdaud the implementation is not failing but there's code redundancy in this step

@Irenyak1 why not log a ticket for it and resolve. Did you get a chance to look at  this? https://opensource.com/article/18/6/anatomy-perfect-pull-request

**Review by kdaud — APPROVED** [review]
cc: @sherrif10

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java [hunk 6]
```diff
@@ -287,7 +287,7 @@ public void endPatientVisit() {
 
 	@Then("the system ends the patient visit")
 	public void systemEndsPatientVisit() {
-		assertNull(visitsDashboardPage.getActiveVisit());
+		assertTrue(textExists("No active visit"));
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java:290]: thought we had changed to use something different? :smiley:

**Review by sherrif10 — APPROVED** [review]
Thanks @Irenyak1 , you guys are consenting again with using presence of texts. Actually using this would be the right approach ideally having this kind of checking the available texts is pretty simple

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java (hunk 6 above)
Comment (Irenyak1) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/ClinicalVisitSteps.java:290]: > thought we had changed to use something different? smiley

@sherrif10  from this [conversation](https://github.com/openmrs/openmrs-contrib-qaframework/pull/256#issuecomment-988907267) and what followed after, seemed like it was agreed to go with texts as opposed to pages.

**Comment by kdaud** [comment]
@dkayiwa can take a final review and perhaps advise accordingly.

**Comment by sherrif10** [comment]
@kdaud  i actually meant this is good to merge

**Comment by Irenyak1** [comment]
> @kdaud i actually meant this is good to merge

@kdaud @sherrif10 @dkayiwa is there a reason why this should not be merged? :-)
